frame/evm/src/lib.rs

```diff
@@ -730,6 +754,24 @@ impl<T: Config> Pallet<T> {
 		}
 
 		<AccountCodes<T>>::insert(address, code);
+
+		// Update metadata.
+		let meta = CodeMetadata::from_code(&code);
+		<AccountCodesMetadata<T>>::insert(address, meta);
+	}
+
+	/// Get the account metadata (hash and size) from storage if it exists,
+	/// or compute it from code and store it if it doesn't exist.
+	pub fn account_code_metadata(address: H160) -> CodeMetadata {
+		if let Some(meta) = <AccountCodesMetadata<T>>::get(&address) {
+			return meta;
+		}
+
+		let code = <AccountCodes<T>>::get(&address);
```

Comment:
It seems to me really unsafe, and not compatible with Weight v2, to force the reading of the code if the metadata are not set.

It would be better to return None and create a dedicated call for migration.

Reply:
> my view is that this is optional. If a chain is upgraded, then they should be recommended to "touch" all contracts upon upgrade (probably through a multi-call to ensure no other code are executed in between)

This is not possible for parachains,  it will make a too big storage proof size. For decentralized parachains, metadata must be computed offchain and injected onchain by a referendum via governance, which requires that the code to support hybrid state.

Reply:
I agree that data should be inserted via governance.

However in the meantime knowing the code hash and size is necessary and require to read the full code. Once it is computed it is better to store it than to have to compute it again the next time. Thus it seems better to not return an Option and keep the unhappy path, which will never be reached once all contracts have been migrated.

Reply:
Ok